fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinLAsscomHelper.java

```diff
@@ -0,0 +1,219 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.doris.nereids.rules.exploration.join;
+
+import org.apache.doris.common.Pair;
+import org.apache.doris.nereids.trees.expressions.Expression;
+import org.apache.doris.nereids.trees.expressions.NamedExpression;
+import org.apache.doris.nereids.trees.expressions.SlotReference;
+import org.apache.doris.nereids.trees.plans.GroupPlan;
+import org.apache.doris.nereids.trees.plans.Plan;
+import org.apache.doris.nereids.trees.plans.logical.LogicalJoin;
+import org.apache.doris.nereids.trees.plans.logical.LogicalProject;
+import org.apache.doris.nereids.util.ExpressionUtils;
+import org.apache.doris.nereids.util.Utils;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Common function for JoinLAsscom
+ */
+public class JoinLAsscomHelper {
+    private final LogicalJoin topJoin;
+    private final LogicalJoin<GroupPlan, GroupPlan> bottomJoin;
+    private final Plan a;
+    private final Plan b;
+    private final Plan c;
+
+    private final Expression topJoinOnClause;
+    private final Expression bottomJoinOnClause;
+
+    private final List<SlotReference> aOutputSlots;
+    private final List<SlotReference> bOutputSlots;
+    private final List<SlotReference> cOutputSlots;
+
+    private final List<Expression> newBottomJoinOnCondition = Lists.newArrayList();
+    private final List<Expression> newTopJoinOnCondition = Lists.newArrayList();
+
+    /**
+     * Init plan and output.
+     */
+    public JoinLAsscomHelper(LogicalJoin topJoin, LogicalJoin<GroupPlan, GroupPlan> bottomJoin) {
```

Comment:
```suggestion
    public JoinLAsscomHelper(LogicalJoin<? extends Plan, GroupPlan> topJoin, LogicalJoin<GroupPlan, GroupPlan> bottomJoin) {
```
then type casts below are not needed